File: 1.Arrays/Q2/Q2.py
"""
Given an array, 
rotate the array to the right by k steps, 
where k is non-negative.
"""

def rotate_array_method_1(array, k):
    if len(array) == k or len(array) == 0:
        return array
    
    counter = 1
    for i in range(k):
        if counter <= k:
            moveItem = array[len(array) - 1]
            del array[-1]
            array.insert(0, moveItem)
            counter += 1
    
    return array

print(rotate_array_method_1([1, 2, 3, 4, 5], 3))

File: 1.Arrays/Q2/Q2.cpp
#include <iostream>
#include <cmath>
#include <algorithm>
using namespace std;

// Given an array, 
// rotate the array to the right by k steps, 
// where k is non-negative.

void rotateArrayMethod1(int array[], int size, int k) {
    k = k % size;
    int temp[k];

    // store last k elements in temp
    for (int i = 0; i < k; i++) {
        temp[i] = array[size - k + i];
    }

    // shift rest of array right by k
    for (int i = size - k - 1; i >= 0; i--) {
        array[i + k] = array[i];
    }

    // copy temp elements to front
    for (int i = 0; i < k; i++) {
        array[i] = temp[i];
    }
}

int main() {
    int arrayMethod1[] = {1, 2, 3, 4, 5};
    int lengthArr = sizeof(arrayMethod1) / sizeof(arrayMethod1[0]);
    int k = 3;

    rotateArrayMethod1(arrayMethod1, lengthArr, k);
    for (int i = 0; i < lengthArr; i++) {
        std::cout << arrayMethod1[i] << " ";
    }
    std::cout << std::endl;
}